stm32 ospi supporting sfdp JESD216

FRASTM opened the pull request:
This PR is adding the support of the read_sfdp and read jedec ID to the stm32 ospi driver.

- The **sdfp-bfp** table is given by the octoFlash controller answering to the READ-SFDP command **or** by the DeviceTree.
If the device-tree of the octoFlash controller has a `sfdp-bfp` table as a property of its DeviceTree, NO read SFDP command is issued to the octoFlash on the bus. The drivers is processing the sdfp table coming from the DeviceTree. 
If NO table is defined as a  `sfdp-bfp` property, the drivers is issuing the  READ-SFDP command to the octoFlash controller.
(in case the READ-SFDP command fails the device cannot be initialized correcly)

The content of the `sfdp-bfp` property of the mx25lm51245 node is copied from the [macronix MX25L51245G datasheet](https://media.digikey.com/pdf/Data%20Sheets/Macronix/MX25L51245G.pdf)  (see 9-46. Read SFDP Mode (RDSFDP))

- The **jedec-id** is given by the octoFlash controller, but is overwritten by the `jedec-id` propery of the device tree (if defined).

The octo-flash of the board is able to display its description with the samples/drivers/jesd216/

According to the zephyr/dts/bindings/mtd/jedec,jesd216.yaml,  if the jedec-id or sfdp table properties are defined by the DTS, it is not read from the octoFlash.: "_This allows encoding the entire BFP block in devicetree to avoid
 reading at runtime, while still allowing the driver to pull out extra data of interest, such as erase sizes._"

Warning :  sfdp-bfp table is given as example in the device tree.

This PR follows the https://github.com/zephyrproject-rtos/zephyr/pull/47744

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/47731

Signed-off-by: Francois Ramu [francois.ramu@st.com](mailto:francois.ramu@st.com)

FRASTM committed:
drivers: flash: ospi for stm32 mcu

Fixes command configuration to have samples/drivers/spi_flash
passed on stm32l562 and stm32u585 disco kits.
In OctoSPI STR/DTR and SPI/STR modes,
to determine the address_width, checking the jesd216_bfp_addrbytes
on JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_3B4B or _VAL_4B.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 ospi driver with read ID from octoflash

This commit adds the jedec216 read JEDEC ID function API.
The ospi commands are issued to the octo flash device or
by the DTS jedec-id property.
Availability of this API is conditional
on selecting CONFIG_FLASH_JESD216_API.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: drivers: jesd216 jedec ID and SFDP from octo-flash

This adds the support jedec configuration to run
the jedec  sample application. So target boards can display
the content of the octo-NOR flash
of the stm32u585 and stm32l562 or stm32h735 disco boards.
The sfdp-bfp table is provided as example (overlay file),
to run this sample on the stm32l562e_dk platform.
It has to be defined in the deviceTree "in cases were runtime
retrieval of SFDP data is not desired."

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32 disco kit with octoflash description

No sfdp-table property given by the DTS but received from
the octoflash Node rely on the issued by the read sfdp command.
Note that the size of the mx25lm51245 flash controller
is expressed in bits (ie 512Mbits or 64 Mbytes).

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
running the samples/drivers/jesd216/ on the b_u585i_iot02a : 
```
*** Booting Zephyr OS build zephyr-v3.1.0-1932-g2e5393ce248d  ***
MX25LM51245: SFDP v 1.6 AP ff with 3 PH
PH0: ff00 rev 1.6: 16 DW @ 30
Summary of BFP content:
DTR Clocking supported
Addressing: 3- or 4-Byte
4-KiBy erase: uniform
Support QSPI XIP
Support 1-1-1
Support 1-1-2: instr 3Bh, 0 mode clocks, 8 waits
Support 1-1-4: instr 6Bh, 0 mode clocks, 8 waits
Support 1-2-2: instr BBh, 0 mode clocks, 4 waits
Support 1-4-4: instr EBh, 2 mode clocks, 4 waits
Support 4-4-4: instr EBh, 2 mode clocks, 4 waits
Flash density: 67108864 bytes
ET1: instr 20h for 4096 By; typ 30 ms, max 420 ms
ET2: instr 52h for 32768 By; typ 160 ms, max 2240 ms
ET3: instr D8h for 65536 By; typ 288 ms, max 4032 ms
Chip erase: typ 59392 ms, max 237568 ms
Byte program: type 32 + 1 * B us, max 128 + 4 * B us
Page program: typ 256 us, max 1024 us
Page size: 256 By
Suspend: B0h ; Resume: 30h
DPD: Enter B9h, exit ABh ; delay 30000 ns ; poll 0x3d
HOLD or RESET Disable: unsupported
QER: 2
0-4-4 Mode methods: entry 0x9 ; exit 0x27
4-4-4 Mode sequences: enable 0x04 ; disable 0xa
4-byte addressing support: enter 0x85, exit 0x3e5
Soft Reset and Rescue Sequence support: 0x10
Status Register 1 support: 0x70
size = <536870912>;
sfdp-bfp = [
        e5 20 fb ff  ff ff ff 1f  44 eb 08 6b  08 3b 04 bb
        fe ff ff ff  ff ff 00 ff  ff ff 44 eb  0c 20 0f 52
        10 d8 00 ff  d6 49 c5 00  81 df 04 e3  44 03 67 38
        30 b0 30 b0  f7 bd d5 5c  4a 9e 29 ff  f0 50 f9 85
        ];
PH1: ffc2 rev 1.0: 4 DW @ 110
sfdp-ffc2 = [
        00 00 00 00  00 00 00 00  10 00 00 00  ff 00 00 00
        ];
PH2: ff84 rev 1.0: 2 DW @ c0
sfdp-ff84 = [
        7f ef ff ff  21 5c dc ff
        ];
jedec-id = [32 31 20];
```

FRASTM commented:
rebase on 60271524087bb4382c8b4db0f2e17a87e4120065
fix CI error

FRASTM commented:
fix CI error
To be rebased once the https://github.com/zephyrproject-rtos/zephyr/pull/46794 is done

GeorgeCGV commented on drivers/flash/jesd216.h:
```diff
@@ -12,6 +12,11 @@
 #include <zephyr/types.h>
 
+/* JEDEC Read identification */
+#define JESD216_CMD_READ_ID   0x9F
```
Duplicates https://github.com/zephyrproject-rtos/zephyr/blob/0514c3b139bffd1813e862a1de42aece0f454c1e/drivers/flash/spi_nor.h#L45

FRASTM replied:
To remove duplicate, it could be 
`#define JESD216_CMD_READ_ID   SPI_NOR_CMD_RDID`

FRASTM commented:
remove JESD216_CMD_READ_ID duplicates in drivers/flash/jesd216.h

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -242,24 +245,79 @@ static OSPI_RegularCmdTypeDef ospi_prepare_cmd(uint8_t transfer_mode, uint8_t tr
+		data[0], data[1], data[2]);
+#endif /* jedec_id */
+	dev_data->cmd_status = 0;
+	id = &data[0];
```
You are assigning `id` pointer to a local variable:
```
	id = &data[0];
```
The `id` correctness is not guaranteed upon exit from the `ospi_read_jedec_id`.

A fix would be to expect the `id` to point to an array of size `JESD216_READ_ID_LEN` outside of the function.

GeorgeCGV reviewed (changes requested):
In the `ospi_read_jedec_id` the `id` pointer points to a local variable allocated on the stack within `ospi_read_jedec_id`:
```
	id = &data[0];
```
The `id` correctness is not guaranteed outside of the `ospi_read_jedec_id` execution.

One fix would be to expect `id` to point to an array of size `JESD216_READ_ID_LEN` outside of the function and read the jedec-id into it.

FRASTM commented:
> In the `ospi_read_jedec_id` the `id` pointer points to a local variable allocated on the stack within `ospi_read_jedec_id`:
> 
> ```
> 	id = &data[0];
> ```
> 
> The `id` correctness is not guaranteed outside of the `ospi_read_jedec_id` execution.
> 
> One fix would be to expect `id` to point to an array of size `JESD216_READ_ID_LEN` outside of the function and read the jedec-id into it.

There is a 	`uint8_t jedec_id[JESD216_READ_ID_LEN];`
where id points : 	`ret = ospi_read_jedec_id(dev, jedec_id);`

FRASTM commented:
The Jedec read ID is hold by a global table filled during init. 
The ospi_read_jedec_id is just reading the table. This avoid two successive  send/receive Commands to the octoFlash.
The ospi_read_jedec_id and ospi_read_sfdp API functions  are condiotnned by the CONFIG_FLASH_JESD216_API.
In case the jedec_id is a property of the DTS

GeorgeCGV commented:
Is there a specific reason to keep `jedec_id` global non `static` instead of having it in the `flash_stm32_ospi_data `?

tobias-aunbol commented:
Just for information, if you are not aware.
I just tested this on my board. And is following hardfault:
<inf> flash_stm32_ospi: OSPI flash config is OPI / DTR
<err> os: ***** USAGE FAULT *****
<err> os:   Division by zero
<err> os: r0/a1:  0x00000001  r1/a2:  0x00000000  r2/a3:  0x20000eb0
<err> os: r3/a4:  0x00000000 r12/ip:  0x0000002b r14/lr:  0x08017e6f
<err> os:  xpsr:  0x69000000
<err> os: Faulting instruction address (r15/pc): 0x08017ebc
<err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
<err> os: Current thread: 0x20004158 (main)
<err> os: Halting system

The faulting address 0x08017ebc is line 1306 in flash_stm32_ospi.c which is:
```
1305	data->layout.pages_size = layout_page_size;
1306	data->layout.pages_count = flash_size / layout_page_size;
1307	LOG_DBG("layout %u x %u By pages", data->layout.pages_count,
1308					   data->layout.pages_size);
```

FRASTM commented:
> Is there a specific reason to keep `jedec_id` global non `static` instead of having it in the `flash_stm32_ospi_data `?

Sure, the jedec_id table could be a part of the flash_stm32_ospi_data (like the write-oc) even this is more a flash "config" (but RO) rather than a flash "data".

FRASTM commented:
For the sdfp table, if the property exists in the DTS, the sfdp_bfp[] is the table of the DTS else the sfdp_bfp[] is filled by reading the octoFlash.

FRASTM commented:
Adding the read_sfdp function to get the parameters either from the table of the DTS if present or from the octoFlash through the READ_SFDP command. For the stm32 target boards, the DTS contains an example of sfdp-bfp table.
The page size is set to a default value in case the info is not available.

FRASTM committed:
drivers: flash: stm32 ospi can read the SFDP table from the DTS

When the SFDP table is provided by the DeviceTree with sfdp-bfp property
It is NOT read from the octoFlash.
If no table exists in the DTS and flash read SPDP fails,
the device init fails.
Availability of this sfdp_read API is conditional
on selecting CONFIG_FLASH_JESD216_API.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Note that giving a board DTS overlay file with : 
```
&mx25lm51245 {
	jedec-id = [aa bb cc];

};
```
will ignore the jedec id read from the octoFlash

GeorgeCGV commented:
> > Is there a specific reason to keep `jedec_id` global non `static` instead of having it in the `flash_stm32_ospi_data `?
> 
> Sure, the jedec_id table could be a part of the flash_stm32_ospi_data (like the write-oc) even this is more a flash "config" (but RO) rather than a flash "data". --> Going to move the table to flash_stm32_ospi_data

I have some code on top that uses read `jedec_id` at runtime to apply certain opcode fixups. That allows usage of different flash ICs for the same board (_not upstreamed as I am not sure if others need that_). However, the id that comes from the DTS (_if any_) resides in config.

erwango commented on boards/arm/b_u585i_iot02a/b_u585i_iot02a-common.dtsi:
```diff
@@ -155,13 +155,6 @@
 		data-rate = <OSPI_DTR_TRANSFER>;
 		four-byte-opcodes;
 		status = "okay";
-		sfdp-bfp = [
```
Can you also fix `size` whichapparently should be size = <DT_SIZE_M(64)*8> ?

FRASTM replied:
converting to Bits

FRASTM commented:
- rebase on 878edeb6ff190a713910a0e5cb9022704c5934a3
- set octoFlash size to 512 Mbits 
- the samples/drivers/jesd216/ running on the b_u585i_iot02a gives the sfdp table directly read in the octoFlash MX25lm51245

Note that running west build -p auto -b b_u585i_iot02a samples/drivers/flash_shell/ with  CONFIG_FLASH_LOG_LEVEL_DBG=y can also display octoFlash parameters

FRASTM commented:
CI failure requires https://github.com/zephyrproject-rtos/zephyr/pull/50300

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -1458,7 +1561,7 @@ static void spi_nor_process_bfp_addrbytes(const struct device *dev,
 	struct flash_stm32_ospi_data *data = dev->data;
 
-	if (jesd216_bfp_addrbytes == JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_4B) {
+	if (jesd216_bfp_addrbytes == JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_3B4B) {
```
Aren't we missing the `JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_4B`?
As 4 bytes are valid for `JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_3B4B` and for `JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_4B`.

FRASTM replied:
yes, JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_4B ORJESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_3B4B

FRASTM commented:
rebase on 9a9fc2b8a5ddf81e44330d9cbebaeba360651e68
add the stm32h735g_disco platform

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -1459,7 +1562,8 @@ static void spi_nor_process_bfp_addrbytes(const struct device *dev,
 	struct flash_stm32_ospi_data *data = dev->data;
 
-	if (jesd216_bfp_addrbytes == JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_4B) {
+	if ((jesd216_bfp_addrbytes == JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_4B) ||
+		(jesd216_bfp_addrbytes == JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_3B4B)) {
```
```suggestion
	if ((jesd216_bfp_addrbytes == JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_4B) ||
	    (jesd216_bfp_addrbytes == JESD216_SFDP_BFP_DW1_ADDRBYTES_VAL_3B4B)) {
```
minor, non-blocking

FRASTM replied:
done

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -116,6 +116,10 @@ struct flash_stm32_ospi_data {
 	enum jesd216_dw15_qer_type qer_type;
+#if defined(CONFIG_FLASH_JESD216_API)
+	/* Table to hold the jedec Read ID given by the octoFlash or the DTS */
+	uint8_t jedec_id[JESD216_READ_ID_LEN]; /* DT_INST_PROP_LEN(0, jedec_id) */
```
```suggestion
	uint8_t jedec_id[JESD216_READ_ID_LEN]; 
```
minor, non-blocking

FRASTM replied:
comment removed

GeorgeCGV reviewed (dismissed):
LGTM
